### DISCUSSION TITLE
MongoDB Database Support

### DISCUSSION THREAD

**Discussion by mrhammadasif** [directus/directus#3156] [Feature Requests]
I think your audience would love to have **MongoDB** support (including me)

**Mongoose** can help with that and as you guys are playing with just plain SQL, I think that would be easy to also support.

Let me consider Advocate of MongoDB. I would love to have discussion on that as well

**Comment by rijkvanzanten** [comment]
I recognize the popularity of MongoDB, and this question has come up before (and will again), but let me ask you this one question:

Why would MongoDB, a document-based data store, be a good solution for a relational datastructure like Directus? 

---

As for your statement 

> Mongoose can help with that and as you guys are playing with just plain SQL, I think that would be easy to also support

While Mongoose can smoothen out some of the unsupported features of Mongo—like nullable constraints, default values, or foreign keys / joins—MongoDB is inherently not made for this purpose, so we'd be adding overhead just to make it act like a relational database.

Also, I'm a little confused by your second point. Because we're "playing with plain SQL" it would be easy to support MongoDB? I don't see how the two are related. 

---

To me, as of right now, supporting MongoDB (or any other non-relational database really) would be mostly for marketing purposes (jumping on the hypetrain), and doesn't (afaict) really solve any problems or offer any benefits to Directus. But please, convince me otherwise 🙂 (That being said, seeing it's so often asked, it's not unthinkable we'll support it in the future 👍 )

**Reply by mrhammadasif** [reply]
While, I somehow agree with your Point of views but as my situation where I want to integrate DirectUS already in the environment where SQL is not allowed, and I need to come up with compatible framework, I have to consider KeystoneJS for this but I am in deep love with the ease and UX of Directus admin panel. which Keystone lacks yet

**Reply by danieljwestman** [reply]
OT, but ”DirectUS” 😂

I’m +1 for **not** putting time on porting Directus to any document based databases. It doesn’t make that much sense, thats my humble opinion in the matter.

**Reply by mrhammadasif** [reply]
Well, in that case, can I have some guidance on how I can create my own database implementation for MongoDB ?

**Reply by rijkvanzanten** [reply]
No, we won't be able to offer that. You're asking me how to overhaul Directus itself by yourself in order to support MongoDB.. That's an absolutely massive task that—as we've discussed above—isn't currently even something that we're looking to support..

**Reply by mrhammadasif** [reply]
are there any plans to support databases plugins to make users able to use their own database solutions?

**Reply by rijkvanzanten** [reply]
Not as of right now, feel free to open a new feature request for that 👍

**Reply by tradaonline** [reply]
Because we can use Mongo Atlas, very good option for cloud services.

**Reply by rijkvanzanten** [reply]
@tradaonline 

> very good option for cloud services.

Care to share _why_ it's a very good option for an app like Directus? See my OP above

**Reply by rvanoord** [reply]
> 
> Why would MongoDB, a document-based data store, be a good solution for a relational datastructure like Directus?
> 

@rijkvanzanten, one motivation would be that MongoDB (or document databases in general) are often a better match for "serverless" / "per request" billing models on cloud platforms like Azure (e.g. via Cosmos DB for MongoDB). In some scenarios this means big cost savings compared to SQL-offerings - which are often tied to a VM under the hood and don't easily scale dynamically.

**Comment by M-GRV** [comment]
https://www.mongodb.com/docs/v7.0/applications/data-models-relationships/

edit: Meant to add a edit tag to comeback to, but was called into a meeting... @rijkvanzanten actually said what was on my mind about a document store.

**Reply by rijkvanzanten** [reply]
How does this relate to the conversation above? Even though it has ways to support it, my original question still stands: _Why would MongoDB, a document-based data store, be a good solution for a relational datastructure?_ 

I'm very interested in potentially supporting MongoDB as a _document_ store, and coming up with a UX that is document-focussed from Directus' perspective, but using MongoDB exclusively as a relational database still doesn't really make sense to me 🤔

**Reply by M-GRV** [reply]
> How does this relate to the conversation above? Even though it has ways to support it, my original question still stands: _Why would MongoDB, a document-based data store, be a good solution for a relational datastructure?_
> 
> I'm very interested in potentially supporting MongoDB as a _document_ store, and coming up with a UX that is document-focussed from Directus' perspective, but using MongoDB exclusively as a relational database still doesn't really make sense to me 🤔

I posted it as a pin for me to comeback to after my meetings for today. I have short term memory loss and it was easier to pin it here and edit my comment later than quickly forgetting and moving on. Sorry for the confusion. 

I was going to post about the potential use of a document store within Directus as well. So you beat me to it while using it's built-in relationship abilities to do so. 

I do agree a relational datastructure being the best way to do it, but there is growing interest in nosql even 3 years after this initial discussion. I have done a pretty big architecture using mongo while using "relational" data in separate collections, but nothing at the scale of what Directus can do. (also relational being a single link back to a user in a user collection)

Another alternative to directus dropped their mongodb support for the same reasons. So again you are not wrong in your thinking. It doesnt fit this use case.

**Reply by rijkvanzanten** [reply]
> I posted it as a pin for me to comeback to after my meetings for today. I have short term memory loss and it was easier to pin it here and edit my comment later than quickly forgetting and moving on. Sorry for the confusion.

Ah! All good 🙂 

> I was going to post about the potential use of a document store within Directus as well. So you beat me to it while using it's built-in relationship abilities to do so.

:) 

> I do agree a relational datastructure being the best way to do it, but there is growing interest in nosql even 3 years after this initial discussion. I have done a pretty big architecture using mongo while using "relational" data in separate collections, but nothing at the scale of what Directus can do. (also relational being a single link back to a user in a user collection)

Surprising statement! I'm obviously in a bit of a relational bubble seeing that's all that we support at the moment, but I've actually been seeing a decline in the amount of times MongoDB shows up in conversations around Directus 🤔 That being said, I have seen an increase in the ability to manage "documents" in some form or fashion (be it a `jsonb` in Postgres or a document database). That's actually a use case we're now actively researching 🙂 My ultimate dream would be to allow you to use SQL for relational data _and_ a document store for document data at the same time. That way, you can leverage the powers of both, without having to deal with the downsides that come with making a relational DB handle documents, or a document database handle relations.

> Another alternative to directus dropped their mongodb support for the same reasons. So again you are not wrong in your thinking. It doesnt fit this use case.

I think the question to me really is: "What would be the ideal use case of a document store within Directus?", rather than "How to make MongoDB work within Directus?", as that opens up way more possibilities and discussion 🙂

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by w0kyj** [comment]
Adding Directus RFC notes and details:
- Reference Issues / Discussions:
    - https://github.com/directus/directus/discussions/3156
    - https://linear.app/directus/issue/ENG-945/implement-basic-mongodb-adapter

# **Summary**

Add support and functionality to manage NOSQL/document-based databases like MongoDB. The support should focus on the key benefits and strengths of a document-based data store. eg:
https://www.mongodb.com/compare/mongodb-mysql

*“MongoDB adopts a significantly different approach to storing data, representing information as a series of JSON-like documents (actually stored as binary JSON, or BSON), as opposed to the table and row format of relational systems.*

*MongoDB documents consist of a series of key/value pairs of varying types, including arrays and nested documents; however, the primary difference is that the structure of the key/value pairs in a given collection can vary from document to document. This more flexible approach is possible because documents are self-describing.”*

**Latest MongoDB Documentation:**
Upcoming: https://www.mongodb.com/docs/v7.0/

Current: https://www.mongodb.com/docs/v6.0/

# **Basic Example**

Notion/Coda style document management

Useful data modeling resource examples:

- https://www.mongodb.com/docs/v7.0/core/data-modeling-introduction/
- https://www.mongodb.com/docs/v7.0/tutorial/model-data-for-atomic-operations/
- https://www.mongodb.com/docs/v7.0/applications/data-models-relationships/
- https://www.mongodb.com/docs/v7.0/tutorial/model-tree-structures-with-parent-references/

# **Motivation**

Expanding the open data platform capabilities of Directus. Providing our clients and users with the ability to choose the best data management solution for the problem/use case being solved. Providing a common data aggregation layer for SQL and NOSQL data sources allows Directus users to derive more value from the platform.

Not all content lends itself to SQL data models and thus support for other data architectures like NOSQL provides a composable and agnostic way for Directus to “democratize” developer design and end user content management.

# **Detailed Design**

Create a data source driver for the new Data Abstraction architecture. May be able to use one of the MongoDB official drivers:

- https://www.mongodb.com/docs/drivers/
    - https://www.mongodb.com/docs/drivers/node/current/
- Alternative library:
    - [https://www.mongodb.com/docs/atlas/partner-integrations/ —>](https://www.mongodb.com/docs/atlas/partner-integrations/) https://mongoosejs.com/

Should we have relational support?

What is a native document relationship?

MongoDB has some relational supported APIs and features (does this map to existing SQL relationships?)

Directus specific constructs (like m2a where it is a code artifact — performance and indexing questions…)

## Requirements List

### Must haves:

- [ ]  API and Configuration support for the following MongoDB options:
    - [ ]  Connect to MongoDB
    - [ ]  Authenticate with MongoDB
    - [ ]  Read from and Write to MongoDB
    - [ ]  Transform your Data (aggregation operations)
    - [ ]  Create and Manage Transactions
    - [ ]  Create Indexes to Speed Up Queries
    - [ ]  Sort Using Collations
    - [ ]  Encrypt Fields from the Client
    - [ ]  Specify BSON Serialization Settings

### Should haves:

- [ ]  Log Events in the Driver
- [ ]  Monitor Driver Events
- [ ]  Run a Database Command

### Could haves:

- [ ]  Store and Retrieve Large Files in MongoDB
- [ ]  Create and Query Time Series Collection
    - [ ]  Put this here since SQL db may be better solution, thus less important for initial driver
- [ ]  Specify Type Parameters with TypeScript

### Will not have:

- [ ]  

# **Drawbacks**

The main drawback will be whether or not the new data abstraction architecture design supports the unstructured data modeling necessary for NOSQL formats. The design must support handling for the unstructured data, indexing for performance and Studio App layouts/interfaces/etc.

# **Alternatives**

n/a

# **Adoption Strategy**

Optional usage and adoption by clients who have a need to manage unstructured data content. No impact to existing clients.

# **Unresolved Questions**

- Studio App visualization of unstructured data
    - Module and/or Layouts
    - Interface(s)
    - Administrative Data Model Management

**Comment by ST9Sergio** [comment]
Hello, as someone who is studying whether to make their backend with directus, one of my biggest concerns is the scalability of the project.

Let's say that it is a fairly ambitious project that, if it works, could "morir de éxito" XD.

MongoDB allows very good horizontal scalability, so having a system like directus for the backend would be spectacular, if not, then I will have to set up a backend project on my own or look for an alternative.

One of my main problems is that Directus does not support multiple databases, and it also does not support schemas that could serve the same purpose.

In such a way that one client of mine is going to affect the performance of the rest of my clients, which is potentially fatal, due to the amount of traffic that the project aims to cover in the long term.

PD: I love everything that directus offers, they have done a great job.

**Reply by pipeninja** [reply]
This. I have been experimenting with Directus and I love it, however the lack of MongoDB support is a bit of a sticking point for me. MongoDB is far easier to run on less powerful hardware and because of this, you can get a lot of performance out of less hardware with MongoDB...and use less power...I totally understand that in certain circumstances MySQL, Postgres et al are probably better solutions...but there are situations where they aren't.

My use case for Directus is to use it as a quick way to throw together a simple, authenticated, backend API that spits out pre-processed data, it's all pretty linear, Directus is very, very good at this but it could be a lot better with MongoDB, I currently do this with SQLite, because I really, really, really don't want to setup a MySQL or Postgres server, my servers here are tuned to draw as little power as possible and as soon as I spin up a SQL database, the energy consumption rises, even if the database isn't experiencing much load, my use cases are "bursty" and don't require huge amounts of processing power or RAM all the time...I recognise that this isn't really inline with the goals for this product, but I bet there are a lot of people using this product for the purpose of rapidly prototyping "linear" APIs...the kind of APIs that process the data on the way in rather than on the way out...this is where document based databases are better...I get that SQL guys think that document based systems are disorganised because they like to neatly arrange data to be retrieved for processing later...and I get that document based guys think that relational databases are an unwieldy mess...because why wait to process it when you can process it now and have the results immediately? Sure you can do that on a relational database, but the retrieval is much slower...if you want an insane firehose of data for tons of users without burning mountains of cash, you need document storage.

**Reply by ST9Sergio** [reply]
I had not thought about it, but if I agree with you, I think that the reason why very few people use MongoDB (at least it happened to Strapi), is simply because SQL is learned first, but for small or large things MongoDB has many advantages because it is much easier to both assemble and scale, both in small and large projects.

I think that even if you use MongoDB as if it were SQL, normalized and with strict rules, it would still be better in most cases, SQL is very powerful, but it usually requires you to be there configuring things and measuring performance.

I think that for a project like Directus, using a database like Mongodb as if it were SQL would be enough, I don't think it is worth maintaining 2 parallel logics for Directus just to support it, if many advantages of MongoDB would be lost, but let's be honest For small or medium projects it is enough and more than enough. For projects that seek to grow in the long term, whatever you set up in directus such as mvp, demo or a section of the business still works.

With SQL I do see that it would be necessary to be able to support multiple databases or schemas, so that you can take advantage of the infrastructure later, if not, then you would have to migrate everything and the truth is that it sounds a bit complex, for when you already have a flow of clients and important data (I don't want to go through that, honestly).

**Comment by weibeu** [comment]
+1

**Comment by Cally99** [comment]
Please do this
